src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs

```diff
@@ -226,22 +226,29 @@ public async Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library
                 }
             }
 
-            while (tasks?.Count > 0)
+            if (tasks?.Count > 0)
             {
-                // Wait for any node to finish resolving
-                var task = await Task.WhenAny(tasks);
-
-                // Extract the resolved node
-                tasks.Remove(task);
-                var dependencyNode = await task;
-                dependencyNode.OuterNode = node;
+                if (tasks.Count > 1)
+                {
+                    // Wait for all the nodes to finish resolving. We only do this if we
+                    // have more than one to avoid WhenAll's defensive copying allocations.
+                    // Otherwise, if there's just one, we'll just end up waiting on it in
+                    // the loop.
+                    await Task.WhenAll(tasks);
+                }
 
-                node.InnerNodes.Add(dependencyNode);
+                foreach (var task in tasks)
+                {
+                    var dependencyNode = await task;
+                    dependencyNode.OuterNode = node;
+                    node.InnerNodes.Add(dependencyNode);
```

Comment:
@nkolev92 This will result in a different order that these dependencies are persisted in InnerNodes; previously it would be have been in fastest to slowest, now its dependency order. Do you know if this is observable?

Reply:
I don't think it should be a problem. 
The walk is exhaustive anyways. If anything, this would make it more deterministic.

Reply:
Resolving this to make it cleaner for other reviewers.